fgpyo/sam/__init__.py

```diff
@@ -1266,8 +1190,220 @@ class SamOrder(enum.Enum):
 
 
 @dataclass(frozen=True)
-class SecondaryAlignment:
-    """A secondary alignment as encoded in one part of the `XA` or `XB` tag value on a SAM record.
+class AuxAlignment(ABC, DataclassInstance):
+    """An alignment parsed from the data stored in the SAM tags of a SAM record.
+
+    Args:
+        reference_name: The reference sequence name.
+        reference_start: The 0-based start position of the alignment.
+        is_forward: If the alignment is in the forward orientation or not.
+        cigar: The Cigar sequence representing the alignment.
+        mapq: The aligner-reported probability of an incorrect mapping, if available.
+        edit_distance: The number of mismatches between the query and the target, if available.
+        alignment_score: The aligner-reported alignment score, if available.
+
+    Raises:
+        ValueError: If reference_start is set to a value less than zero.
+        ValueError: If mapq is set to a value less than zero.
+        ValueError: If edit_distance is set to a value less than zero.
+        ValueError: If alignment_score is set to a value less than zero.
+    """
+
+    sam_tags: ClassVar[Collection[str]]
+    """The SAM tags this class is capable of parsing."""
+
+    reference_name: str
+    reference_start: int
+    is_forward: bool
+    cigar: Cigar
+    mapq: Optional[int] = None
+    edit_distance: Optional[int] = None
+    alignment_score: Optional[int] = None
+
+    def __post_init__(self) -> None:
+        """Perform post-initialization validation on this dataclass."""
+        errors: list[str] = []
+        if self.reference_start < 0:
+            errors.append(f"Reference start cannot be less than 0! Found: {self.reference_start}")
+        if self.mapq is not None and self.mapq < 0:
+            errors.append(f"Mapping quality cannot be less than 0! Found: {self.mapq}")
+        if self.edit_distance < 0:
+            errors.append(f"Edit distance cannot be less than 0! Found: {self.edit_distance}")
+        if self.alignment_score is not None and self.alignment_score < 0:
+            errors.append(f"Alignment score cannot be less than 0! Found: {self.alignment_score}")
```

Comment:
why not?  I think a negative alignment score is aligned, especially in a global alignment.

Reply:
I'd like to learn more about this concern! I haven't seen a score below zero for any output from `bwa` so I figured it would be a sign that data was corrupt. Couple questions:

1. Is there a way to run `bwa` in which the score can go below zero?
2. Knowing that this code will probably parse `bwa` tags for the forseeable future, do you think keeping the strict check is better for safety-sake or would hinder someone in the future with a popular or custom global aligner putting auxiliary alignments in SAM tags?